From 9bb7a7a45d5325eae5ec8bed13157f7e65d0bfea Mon Sep 17 00:00:00 2001
From: Sebastian Pipping <sebastian@pipping.org>
Date: Mon, 26 Oct 2026 17:17:57 +0100
Subject: [PATCH] tests: Add a test for a initial non-final empty subparser
 parse

.. with a handler call
[Hades: from libexpat pull request #1444, commit 9bb7a7a45d5325eae5ec8bed13157f7e65d0bfea,
 not merged as of 2026-10-28; taken for CVE-2026-117610 and CVE-2026-118370.]

---
 tests/misc_tests.c | 63 ++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 63 insertions(+)

diff --git a/tests/misc_tests.c b/tests/misc_tests.c
index 5d87c92..6bfde9b 100644
--- a/tests/misc_tests.c
+++ b/tests/misc_tests.c
@@ -1064,6 +1064,68 @@ START_TEST(test_misc_start_doctype_decl_handler_reset) {
 }
 END_TEST
 
+static void
+comparingPiHandler(void *userData, const XML_Char *target,
+                   const XML_Char *data) {
+  UNUSED_P(userData);
+
+  XML_Char *const literalTarget = xcsdup_str("target");
+  XML_Char *const literalData = xcsdup_str("data");
+  assert_true(literalTarget != NULL);
+  assert_true(literalData != NULL);
+
+  assert_true(xcscmp(target, literalTarget) == 0);
+  assert_true(xcscmp(data, literalData) == 0);
+
+  free(literalTarget);
+  free(literalData);
+}
+
+static int XMLCALL
+piParsingExternalEntityHandler(XML_Parser parent, const XML_Char *context,
+                               const XML_Char *base, const XML_Char *systemId,
+                               const XML_Char *publicId) {
+  UNUSED_P(base);
+  UNUSED_P(systemId);
+  UNUSED_P(publicId);
+
+  const char *const doc = "<?target data?>";
+
+  XML_Parser subParser = XML_ExternalEntityParserCreate(parent, context, NULL);
+  assert_true(subParser != NULL);
+
+  // NOTE: Related test "test_empty_ext_param_entity_in_value" could be of
+  // interest
+  assert_true(_XML_Parse_SINGLE_BYTES(subParser, "", 0, /*isFinal=*/XML_FALSE)
+              == XML_STATUS_OK);
+  assert_true(_XML_Parse_SINGLE_BYTES(subParser, doc, (int)strlen(doc),
+                                      /*isFinal=*/XML_TRUE)
+              == XML_STATUS_OK);
+
+  XML_ParserFree(subParser);
+
+  return XML_STATUS_OK;
+}
+
+START_TEST(test_misc_empty_nonfinal_subparser_parse) {
+  const char *const doc = "<!DOCTYPE root [\n"
+                          "  <!ENTITY ext SYSTEM 'external.dtd'>\n"
+                          "]>\n"
+                          "<root>&ext;</root>\n";
+
+  XML_Parser parser = XML_ParserCreate(NULL);
+  assert_true(parser != NULL);
+
+  XML_SetProcessingInstructionHandler(parser, comparingPiHandler);
+  XML_SetExternalEntityRefHandler(parser, piParsingExternalEntityHandler);
+
+  assert_true(_XML_Parse_SINGLE_BYTES(parser, doc, (int)strlen(doc),
+                                      /*isFinal=*/XML_TRUE));
+
+  XML_ParserFree(parser);
+}
+END_TEST
+
 void
 make_miscellaneous_test_case(Suite *s) {
   TCase *tc_misc = tcase_create("miscellaneous tests");
@@ -1105,4 +1167,5 @@ make_miscellaneous_test_case(Suite *s) {
   tcase_add_test(tc_misc, test_misc_input_2gb);
   tcase_add_test(tc_misc, test_misc_low_surrogate_mozilla_bug_2053153);
   tcase_add_test(tc_misc, test_misc_start_doctype_decl_handler_reset);
+  tcase_add_test(tc_misc, test_misc_empty_nonfinal_subparser_parse);
 }
-- 
2.53.0

